application/src/main/java/org/togetherjava/tjbot/commands/TopHelpers.java

```diff
@@ -0,0 +1,114 @@
+
+public final class TopHelpers implements Command {
+    private static final Logger logger = LoggerFactory.getLogger(TopHelpers.class);
+
+    private static final String COUNT_OPTION = "count";
+    private static final String NO_ENTRIES = "No entries";
+
+    private static final long MIN_COUNT = 10L;
+    private static final long MAX_COUNT = 30L;
+
+    private final Database database;
+
+    /**
+     * Initializes TopHelpers with a database.
+     *
+     * @param database the database to store the key-value pairs in
+     */
+    public TopHelpers(Database database) {
+        this.database = database;
+    }
+
+    @Override
+    public @NotNull String getCommandName() {
+        return "tophelpers";
+    }
+
+    @Override
+    public @NotNull String getDescription() {
+        return "Find top helpers for last 30 days";
+    }
+
+    @Override
+    public boolean isGuildOnly() {
+        return true;
+    }
+
+    @Override
+    public @NotNull CommandData addOptions(@NotNull CommandData commandData) {
+        return commandData.addOptions(new OptionData(OptionType.STRING, COUNT_OPTION,
+                "Count of top helpers to be retrieved (default is 10 and capped at 30)", false));
+    }
+
+    @Override
+    public void onSlashCommand(SlashCommandEvent event) {
+        var guildId = event.getGuild().getIdLong();
+        var count = getBoundedCount(event.getOption(COUNT_OPTION));
+        database.read(dsl -> {
+            var records = dsl.with("TOPHELPERS")
+                .as(select(MESSAGE_METADATA.USER_ID, count().as("COUNT")).from(MESSAGE_METADATA)
+                    .where(MESSAGE_METADATA.GUILD_ID.eq(guildId))
+                    .groupBy(MESSAGE_METADATA.USER_ID)
+                    .orderBy(count().desc())
+                    .limit(count))
+                .select(rowNumber().over(orderBy(field(name("COUNT")).desc())).as("#"),
+                        field(name("USER_ID"), Long.class), field(name("COUNT"), Long.class))
+                .from(table(name("TOPHELPERS")))
+                .fetch();
+            event.reply(prettyPrintOutput(buildTopHelpersDataFrame(event, records))).queue();
+        });
+    }
+
+    private long getBoundedCount(OptionMapping count) {
+        return count == null ? MIN_COUNT : Math.min(count.getAsLong(), MAX_COUNT);
+    }
+
+    private List<List<String>> buildTopHelpersDataFrame(SlashCommandEvent event,
+            Result<Record3<Integer, Long, Long>> records) {
+        return Arrays.stream(records.intoArrays())
+            .map(objects -> List.of(objects[0].toString(),
+                    getMemberNameFromEvent(event, objects[1]), objects[2].toString()))
+            .toList();
+    }
+
+    private String getMemberNameFromEvent(SlashCommandEvent event, Object member) {
+        String memberName;
+        try {
+            memberName = event.getGuild().getMemberById((Long) member).getEffectiveName();
```

Comment:
Use user instead, always works when the ID exists

Reply:
I assume that's another API but the check here was for if the userId doesnt exists. So how does that help?

Reply:
Guild#getMemberById, like the docs say, only check cached members.
[See the default caching policy](https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/utils/MemberCachePolicy.html#DEFAULT)
And, people that left the guild display as [UNKNOWN MEMBER] now.

So instead, JDA#retreiveUserById solves both of those.
Only thing required is a valid account.

Yes you lose effective nicknames, so if you really want that you're forced to retrieve the user and member, or to cache all users.
Both not really ideal, but regardless I'd recommend retrieving myself.

Reply:
Yeah looks like I have to make it blocking to maintain simplicity of the codebase.

Reply:
Then write it differently, async can still be simple.
All you do is copy paste the logic into a lambda when done correctly.
Guild#retrieveMembersByIds is what you should use instead

```java
Collection<Long> memberIds = List.of(1000L, 200L);
guild.retrieveMembersByIds(memberIds).onSuccess(list -> {
    Map<Long, String> memberNicknameMap = list.stream().collect(Collectors.toMap(Member::getIdLong, Member::getEffectiveName));
    
    memberIds.forEach(memberId -> {
        System.out.println(memberNicknameMap.get(memberId));
    });
```

Already does everything for you, you just need to loop through all members.
You get their nickname, and format it to however you want.

And after further notice, I just noticed Discord has no retrieve users endpoint so you can just forget about that honestly.

Reply:
No, that's a consumer all I want is a `RestAction<User>` or a `RestAction<String>` this way I can extract logical blocks into specific methods. Fact is JDA type system isn't rich enough to allow for this. It's a collision of wannabe reactive RestAction with CompletableFuture as far as I can see

Reply:
There's map and flatmap, why cant you use those? It works literally the same as your reactive libraries, just less methods.
And retrieving users one by one is pretty bad, so I'd say retrieveMembersById

Reply:
1. You are jumping back and forth between your suggestions. retrieveMembersById is again for cached members only
2. There is nothing about `map/flatMap` in this discussion. The reactive type RestAction is only bound to well REST actions ie. web endpoints and one you get out of that world you need to figure out to map those to other types that Java has. Just like your wouldn't mix and match nio with io(which you can but you shouldnt ideally) or streams with loops

There isn't a reactive Producer type that you can use across your API's which is bad. It's a mismatch of sorts.

Reply:
How is retrieveMembersByIds only for cached users?
Did you even read the docs?

Yes if you're using retrieveMembersByIds there's no RestAction. You mentioned a RestAction<User> and that implies you're retrieving users manually.

Still, I showed you a simple example of how it can be done

Don't make it more complicated then it is.

Yes there's no reactor like thing, if you want reactive programming this is the wrong library, this just allows mapping.

Reply:
Apologies probably my IDE autocompleted but all of getMemberByXXX methods of Guild retrieve data from cache.
![image](https://user-images.githubusercontent.com/3216900/133382544-12ba6114-81a8-4201-b328-15c190807768.png)

Reply:
Where did I mention getMembersByIds
![image](https://user-images.githubusercontent.com/49957334/133382710-05e2c5d2-41ed-4238-8f9a-dc454f8f367b.png)
![image](https://user-images.githubusercontent.com/49957334/133382719-a505aee0-68f3-428c-b966-053d2e8e3c54.png)
![image](https://user-images.githubusercontent.com/49957334/133382748-cefc635d-0028-429a-8bd2-66b461b90918.png)

Reply:
My bad. But now it has a new type Task lol!!

Reply:
I gave you code above, in how you can do it, assuming you're alright with using consumers.

Reply:
And yeah, it's a Task not a RestAction.
Why? Because it's a gateway request, not a rest request.